common/cache/budget.go

```diff
@@ -0,0 +1,1260 @@
+			fairShareAvailableCount = fairSharePerCacheCount - currentFairShareCount
+		}
+
+		// Total soft cap available is free space + fair share for this cache
+		softCapAvailableBytes := freeSpaceBytes + fairShareAvailableBytes
+		softCapAvailableCount := freeSpaceCount + fairShareAvailableCount
+
+		// Available capacity is minimum of soft cap and hard cap constraints
+		availableBytes := min(softCapAvailableBytes, hardCapAvailableBytes)
+		availableCount := min(softCapAvailableCount, hardCapAvailableCount)
+
+		// Check if the request can be satisfied
+		if additionalBytes <= availableBytes && additionalCount <= availableCount {
+			// Success - calculate allocation breakdown
+			freeBytesUsage := min(additionalBytes, freeSpaceBytes)
+			fairShareBytesUsage := additionalBytes - freeBytesUsage
+			freeCountUsage := min(additionalCount, freeSpaceCount)
+			fairShareCountUsage := additionalCount - freeCountUsage
+
+			return CapEnforcementResult{
+				Error:          nil,
+				FreeBytes:      freeBytesUsage,
+				FairShareBytes: fairShareBytesUsage,
+				FreeCount:      freeCountUsage,
+				FairShareCount: fairShareCountUsage,
+				AvailableBytes: availableBytes,
+				AvailableCount: availableCount,
+			}
+		}
+
+		// Determine which constraint failed
+		var err error
+		if additionalBytes > hardCapAvailableBytes || additionalCount > hardCapAvailableCount {
+			// Hard cap constraint failed
+			if additionalBytes > hardCapAvailableBytes {
+				err = ErrBytesBudgetExceeded
+			} else {
+				err = ErrCountBudgetExceeded
+			}
+		} else {
+			// Soft cap constraint failed
+			if additionalBytes > softCapAvailableBytes {
+				err = ErrBytesSoftCapExceeded
+			} else {
+				err = ErrCountSoftCapExceeded
+			}
+		}
+
+		return CapEnforcementResult{
+			Error:          err,
+			FreeBytes:      0, // No allocation on failure
+			FairShareBytes: 0, // No allocation on failure
+			FreeCount:      0, // No allocation on failure
+			FairShareCount: 0, // No allocation on failure
+			AvailableBytes: availableBytes,
+			AvailableCount: availableCount,
+		}
+	}
+
+	// Soft caps disabled - only check hard cap
+	availableBytes := hardCapAvailableBytes
+	availableCount := hardCapAvailableCount
+
+	if additionalBytes <= hardCapAvailableBytes && additionalCount <= hardCapAvailableCount {
+		// Success - all allocation goes to fair share when soft caps disabled
+		return CapEnforcementResult{
+			Error:          nil,
+			FreeBytes:      0, // No free space allocation when soft caps disabled
+			FairShareBytes: additionalBytes,
+			FreeCount:      0, // No free space allocation when soft caps disabled
+			FairShareCount: additionalCount,
+			AvailableBytes: availableBytes,
+			AvailableCount: availableCount,
+		}
+	}
+
+	// Hard cap exceeded
+	var err error
+	if additionalBytes > hardCapAvailableBytes {
+		err = ErrBytesBudgetExceeded
+	} else {
+		err = ErrCountBudgetExceeded
+	}
+
+	return CapEnforcementResult{
+		Error:          err,
```

Comment:
Do you mind removing this and pulling it to it's own error type as with the other one?